ipaserver/plugins/dogtag.py

```diff
@@ -1100,7 +1100,15 @@ def parse_updateCRL_xml(doc):
 #-------------------------------------------------------------------------------
 
 from ipalib import Registry, errors, SkipPluginModule
-if api.isdone('finalize') and api.env.ra_plugin != 'dogtag':
+if api.isdone("finalize") and not (
+    api.env.ra_plugin == "dogtag"
+    or api.env.context
+    in (
+        "installer",
+        "server",
+        "lite",
+    )
+):
```

Comment:
I am puzzled by this change and fail to understand why it is needed. When the context is installer, server or lite the plugin is always loaded, even if ra_plugin is None. In which situation is it used? Is it because during promotion the ra_plugin is initially None (because we start from a client-only) ?
I need to confess that all the variables in the installer vs the api are confusing for me: ra_only, clone, promote / ra_plugin, enable_ra.

Reply:
The bulk of the dogtag module is not loaded unless certain conditions are met, making the RA interface unavailable. The RA interface is needed to load the ACME profile into the CA.

The root of the problem is that during promotion ra_plugin is None. Trying to set it early caused a whole other series of hard to debug problems so I went with context. I initially tried using _only_ context but that broke upgrades. Given there could be other unknowns in the current patch I try the least-possible change. I should limit the context to only installer since I am no longer dropping the ra_plugin test.

ra_plugin and enable_ra date back to when there was a lot more abstraction around the CA since we supported more than just dogtag. There were originally two ra_plugins, dogtag and selfsign. So except for a CA-less install, any IPA server can communicate with the CA but only some have the CA installed (so they are all RA's but only some are CA's). enable_ra probably made more sense in the past with the abstraction layer.

ra_only is completely confusing and seems to really mean "setup-ca was passed in." 

IIRC at one point a fair bit of code was duplicated in order to have two separate paths for initial installation and replica install. These were merged and then DL1 (promoting a client to server) was added which added additional complexity.